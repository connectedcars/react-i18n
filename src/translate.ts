const DEFAULT_LANG = 'en'
const DEFAULT_PLURAL = 'nplurals=2; plural=(n != 1);'
const CONTEXT_GLUE = '\u0004'

export type TranslationBlock = string | null

export interface TranslationSet {
  [msgid: string]: TranslationBlock[]
}

export interface Translations {
  [lang: string]: TranslationSet
}

export const getTranslation = (translations: Translations, lang: string) => (
  n: number,
  text: string,
  textPlural?: string | null,
  context?: string | null
): string => {
  lang = lang || DEFAULT_LANG

  const fallback = getFallbackTranslation(n, text, textPlural)

  if (lang === DEFAULT_LANG) {
    return fallback
  }

  // Make sure the language exists in our translation set.
  if (translations.hasOwnProperty(lang) === false) {
    console.warn(`translations does not contain language: '${lang}'`)
    return fallback
  }

  // Get our language set.
  const translationSet = translations[lang]

  // Generate our message id. If we have a context, join them.
  const msgid = context != null
    ? context + CONTEXT_GLUE + text
    : text

  // Get the translation values.
  // First entry is our `textPlural` and the following are translations.
  const msgstr = (translationSet[msgid] || []).slice()
  if (msgstr.length === 0) {
    console.warn(`translation not found for: text='${text}', text_plural='${textPlural}'`)
    return fallback
  }

  // Shift off the first entry, it's our original translation plural.
  const msgstrPlural = msgstr.shift()

  // Deal with plurals.
  if (textPlural) {
    if (textPlural !== msgstrPlural) {
      console.warn(`translations for textPlural does not match '${textPlural}' != '${msgstrPlural}'`)
      return fallback
    }

    // Get plural-forms from the header.
    let pluralForms = translationSet[''] && translationSet['']['plural-forms']

    if (!pluralForms) {
      console.warn(
        `translations does not have a plural-forms setting on language: '${lang}', default to: '${DEFAULT_PLURAL}'`
      )
      pluralForms = DEFAULT_PLURAL
    }

    const msgstrIndex = pluralFunc(pluralForms)(n)
    return msgstr[msgstrIndex] || fallback
  }

  // Deal with singulars.
  return msgstr[0] || fallback
}

const getFallbackTranslation = (n: number, text: string, textPlural?: string | null) => {
  if (textPlural != null) {
    return [text, textPlural][pluralFunc(DEFAULT_PLURAL)(n)]
  }
  return text
}

export const pluralFunc = (pluralForms: string) => {
  const code = [
    'var plural;',
    'var nplurals;',
    pluralForms,
    'return (plural === true ? 1 : plural ? plural : 0);'
  ]
  return Function('n', code.join('\n'))
}

export const replace = (text: string, count: number, data?: Object) => {
  let ourData: any = data || {}
  if (ourData.hasOwnProperty('n')) {
    console.warn('translations data keyword `n` is reserved')
  }
  ourData.n = count
  Object.keys(ourData).map(key => {
    text = text.replace(new RegExp(`{${key}}`, 'g'), ourData[key])
  })
  return text
}
