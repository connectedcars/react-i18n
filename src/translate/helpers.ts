import { FormatLocaleOption } from '../types'

export const parseLocale = (locale: string) => {
  const [languageCode, countryCode] = locale.split(/[-_]/)
  return {
    languageCode,
    countryCode: countryCode?.toUpperCase() || null,
  }
}

export const formatLocale = (locale: string, format: FormatLocaleOption) => {
  const { languageCode, countryCode } = parseLocale(locale)
  if (!countryCode) {
    return languageCode.toLowerCase()
  }
  return format
    .replace('xx', languageCode.toLowerCase())
    .replace('XX', languageCode.toUpperCase())
    .replace('yy', countryCode.toLowerCase())
    .replace('YY', countryCode.toUpperCase())
}

export const getSupportedLocaleFromLocalesList = (
  clientLocales: readonly string[],
  translationLocales: string[],
  format: FormatLocaleOption = 'xx_YY'
): string | null => {
  const locales = clientLocales.map((v) => {
    return formatLocale(v, format)
  })

  // Filter all matching languages on the first part of the locale.
  const supported = translationLocales
    .filter((l1) => {
      const [userLocale] = l1.split(/[-_]/)
      return locales.find((l2) => {
        const [locale] = l2.split(/[-_]/)
        return locale.toLowerCase() === userLocale.toLowerCase()
      })
    })
    .map((v) => formatLocale(v, format))

  // Attempt to find direct matches first
  for (const l of locales) {
    if (supported.includes(l)) {
      return l
    }
  }

  // If we can't find a direct match, try matching the first part of the locale.
  for (const l of locales) {
    const found = supported.find(
      (x) => x.split(/[-_]/)[0] === l.split(/[-_]/)[0]
    )
    if (found) {
      return found
    }
  }

  return null
}
