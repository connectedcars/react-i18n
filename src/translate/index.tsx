export { mergeTranslations } from './merge-translations'
export { normalizeContent } from './normalize-content'
export { replaceJsx, replaceString } from './replace-content'
export { getTranslation } from './translate'
