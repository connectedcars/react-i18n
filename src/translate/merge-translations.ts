import { Translations } from "../types"

export const mergeTranslations = (translations: Translations[]): Translations => {
  const mergedTranslations: Translations = {}

  for (const t of translations) {
    for (const key of Object.keys(t)) {
      const value = t[key]
      if (mergedTranslations[key]) {
        // This is a naive implementation. This overrides translations in previous sets, so the last one will always be the one picked.
        // Even if something is translated in one translation file, but isn't in the last file, they will still be replaced.
        mergedTranslations[key] = {
          ...mergedTranslations[key],
          ...value
        }
      } else {
        mergedTranslations[key] = value
      }
    }
  }

  return mergedTranslations
}
