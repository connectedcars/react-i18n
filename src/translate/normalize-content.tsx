import { TranslationOptions } from '../types'

export const normalizeContent = (
  message: string,
  options?: TranslationOptions['content']
) => {
  const {
    trimWhiteSpace = true,
    preserveIndentation = false,
    replaceNewLines = false,
  } = options || {}

  if (trimWhiteSpace) {
    message = message.replace(/^\n+|\s+$/g, '')
  }
  if (!preserveIndentation) {
    message = message.replace(/^[ \t]+/gm, '')
  }
  message = message.replace(/\r\n/g, '\n')
  if (typeof replaceNewLines === 'string') {
    message = message.replace(/^\n+|\s+$/g, replaceNewLines)
  }

  return message
}
