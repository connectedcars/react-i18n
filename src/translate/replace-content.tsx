import React from 'react'

import { DocNode, ElementNode, parse, TextNode } from '../parser'
import {
  ReplaceJsxOptions,
  ReplaceStringOptions,
  ReplaceStringRegex,
  TranslateDataWithJSX,
} from '../types'

const defaultReplaceStringRegex: ReplaceStringRegex = {
  pattern: (key) => `{${key}}`,
}

export const replaceString = (
  str: string,
  data: TranslateDataWithJSX | null,
  options?: ReplaceStringOptions
) => {
  const { strict = false, replaceStringRegex = defaultReplaceStringRegex } =
    options || {}
  if (!data) {
    return str
  }

  Object.keys(data).forEach((key) => {
    const v = data[key]
    if (typeof v === 'function') {
      return
    }
    str = str.replace(
      new RegExp(replaceStringRegex.pattern(key), 'g'),
      String(v)
    )
  })

  if (strict) {
    const match = str.match(new RegExp(replaceStringRegex.pattern('([\\w-]+)')))
    if (match) {
      throw new Error(`translation data not found for tag: '${match[1]}'`)
    }
  }

  return str
}

export const replaceJsx = (
  str: string,
  data: TranslateDataWithJSX | null,
  options?: ReplaceJsxOptions
): React.ReactNode[] => {
  const { strict = false, replaceStringRegex = defaultReplaceStringRegex } =
    options || {}

  const keys = Object.keys(data || {})
  if (strict) {
    keys.push('[\\w-]+')
  }
  return parse(str, replaceStringRegex.pattern(`(${keys.join('|')})`)).map(
    (node) => renderNode(node, data, strict)
  )
}

const renderNode = (
  node: DocNode,
  data: TranslateDataWithJSX | null,
  strict: boolean
): React.ReactNode => {
  if (node instanceof TextNode) {
    return node.text
  }

  if (node instanceof ElementNode) {
    if (data) {
      const dataValue = data[node.tagName]
      if (dataValue != null) {
        if (typeof dataValue === 'function') {
          return dataValue(
            node.children.map((node) => renderNode(node, data, strict)),
            node.attributes
          )
        }

        return dataValue
      }
    }

    if (strict) {
      throw new Error(`translation data not found for tag: '${node.tagName}'`)
    }

    return React.createElement(
      node.tagName,
      { key: node.text },
      node.children.map((node) => renderNode(node, data, strict))
    )
  }

  throw new Error(
    `translation node type is not supported: ${JSON.stringify(node)}`
  )
}
