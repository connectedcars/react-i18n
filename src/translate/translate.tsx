import { TranslationSet, TranslationOptions } from '../types'
import { normalizeContent } from './normalize-content'

const CONTEXT_GLUE = '\u0004'

export const getTranslation = (
  translationSet: TranslationSet | undefined,
  n: number | null,
  singular: string,
  plural?: string | null,
  context?: string | null,
  options?: TranslationOptions
): string => {
  // Clean our translation.
  singular = normalizeContent(singular, options.content)

  const defaultValue = plural ? (n !== 1 ? plural : singular) : singular

  if (!translationSet) {
    return defaultValue
  }

  // Generate our message id. If we have a context, join them.
  const msgid = context != null ? context + CONTEXT_GLUE + singular : singular

  // Get the translation values.
  // First entry is our `plural` and the following strings are translations.
  const msgstr = (translationSet[msgid] || []).slice()
  if (msgstr.length === 0) {
    console.warn(
      `translation not found for: msg='${singular}', msg_plural='${plural}'`
    )
    return defaultValue
  }

  // Shift off the first entry, it's our original translation plural.
  const msgstrPlural = msgstr.shift()

  // Deal with singulars.
  if (!plural) {
    return msgstr[0] || defaultValue
  }

  // Deal with plurals.
  if (plural !== msgstrPlural) {
    console.warn(
      `translations for plural does not match '${plural}' != '${msgstrPlural}'`
    )
    return defaultValue
  }

  // Get plural-forms from the header.
  let pluralForms = translationSet[''] && translationSet['']['plural-forms']

  if (!pluralForms) {
    console.warn(`translations are missing plural-forms setting`)
    return defaultValue
  }

  const msgstrIndex = getPluralFunc(pluralForms as string)(n)

  return msgstr[msgstrIndex] || defaultValue
}

const getPluralFunc = (pluralForms: string) => {
  const code = [
    'var plural;',
    'var nplurals;',
    pluralForms,
    'return (plural === true ? 1 : plural ? plural : 0);',
  ]
  return Function('n', code.join('\n'))
}
