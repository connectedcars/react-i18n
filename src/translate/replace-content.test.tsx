import React from 'react'
import { replaceString } from '../translate'
import { replaceJsx } from './replace-content'

describe('replace-content', () => {
  it('can replace a string', () => {
    expect(
      replaceString('Hello {name}', { name: 'John Doe' })
    ).toMatchSnapshot()
  })

  it('can replace JSX', () => {
    expect(
      replaceJsx('Hello <strong>{name}</strong>', { name: 'John Doe' }, false)
    ).toMatchSnapshot()

    expect(() =>
      replaceJsx('Hello <strong>{name}</strong>', { name: 'John Doe' }, true)
    ).toThrowErrorMatchingSnapshot()

    const Link: React.FC = ({ children }) => {
      return <a href="https://example.com">{children}</a>
    }

    expect(
      replaceJsx('<strong>Hello</strong> World <link>Foo bar</link>', {
        strong: content => <strong>{content}</strong>,
        link: content => <Link>Link here</Link>
      }, true)
    ).toMatchSnapshot()
  })
})
