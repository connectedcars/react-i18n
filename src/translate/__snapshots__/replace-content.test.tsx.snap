// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replace-content can replace JSX 1`] = `
[
  "Hello ",
  <strong>
    John Doe
  </strong>,
]
`;

exports[`replace-content can replace JSX 2`] = `"translation data not found for tag: 'strong'"`;

exports[`replace-content can replace JSX 3`] = `
[
  <strong>
    Hello
  </strong>,
  " World ",
  <Link>
    Link here
  </Link>,
]
`;

exports[`replace-content can replace a string 1`] = `"Hello John Doe"`;

exports[`replace-content can replace a string with a custom pattern 1`] = `"Hello John Doe"`;

exports[`replace-content can replace a string with a custom pattern 2`] = `"Hello John Doe"`;
